packages/core/src/browser/frontend-connection-status.ts

```diff
@@ -0,0 +1,217 @@
+                const { readyState, status } = xhr;
+                if (readyState === XMLHttpRequest.DONE) {
+                    handle(status === 200);
+                }
+            };
+            xhr.onerror = () => handle(false);
+            xhr.ontimeout = () => handle(false);
+            xhr.open('GET', `${this.endpointUrl}/alive`);
+            try {
+                xhr.send();
+            } catch {
+                handle(false);
+            }
+        });
+    }
+
+    protected setTimeout(handler: (...args: any[]) => void, timeout: number): number {
+        return window.setTimeout(handler, timeout);
+    }
+
+    protected clearTimeout(handle: number): void {
+        window.clearTimeout(handle);
+    }
+
+}
+
+@injectable()
+export class ConnectionStatusStatusBarContribution extends DefaultFrontendApplicationContribution {
+
+    constructor(
+        @inject(ConnectionStatusService) protected readonly connectionStatusService: ConnectionStatusService,
+        @inject(StatusBar) protected statusBar: StatusBar
+    ) {
+        super();
+        this.connectionStatusService.onStatusChange(status => this.onStatusChange(status));
+    }
+
+    protected onStatusChange(status: ConnectionStatus) {
+        this.statusBar.removeElement('connection-status');
+        const text = `$(${this.getStatusIcon(status)})`;
+        const tooltip = this.getStatusTooltip(status);
+        this.statusBar.setElement('connection-status', {
+            alignment: StatusBarAlignment.RIGHT,
+            text,
+            priority: 0,
+            tooltip
+        });
+    }
+
+    protected getStatusIcon(status: ConnectionStatus) {
+        if (status.state === ConnectionState.IDLE) {
+            return 'bed';
+        }
+        const { health } = status;
+        if (health === undefined || health === 0) {
+            return 'exclamation-circle';
+        }
+        if (health < 25) {
+            return 'frown-o';
+        }
+        if (health < 50) {
+            return 'meh-o';
+        }
+        return 'smile-o';
+    }
+
+    protected getStatusTooltip(status: ConnectionStatus) {
+        if (status.state === ConnectionState.IDLE) {
+            return 'Idle';
+        }
+        return status.health ? `Online [Connection health: ${status.health}%]` : 'Offline';
+    }
+
+}
+
+@injectable()
+export class ApplicationConnectionStatusContribution extends DefaultFrontendApplicationContribution {
+
+    private dialog: ConnectionStatusDialog | undefined;
+    private state = ConnectionState.ONLINE;
+
+    constructor(
+        @inject(ConnectionStatusService) protected readonly connectionStatusService: ConnectionStatusService,
+        @inject(ILogger) protected readonly logger: ILogger
+    ) {
+        super();
+        this.connectionStatusService.onStatusChange(status => this.onStatusChange(status));
+    }
+
+    protected onStatusChange(status: ConnectionStatus): void {
+        if (this.state !== status.state) {
```

Comment:
ok, have not thought about health rate

Reply:
I will change this to have to individual hooks. One for the state and one for the status changes.